Sources/Workspace/Workspace+BinaryArtifacts.swift

```diff
@@ -459,6 +473,56 @@ extension Workspace.BinaryArtifactsManager {
     }
 }
 
+extension Workspace.BinaryArtifactsManager {
+
+    static func deriveBinaryArtifact(fileSystem: FileSystem, path: AbsolutePath) throws -> (AbsolutePath, BinaryTarget.Kind)? {
+        guard fileSystem.exists(path) else {
+            return .none
+        }
+
+        // is the current path it?
+        if let kind = try deriveBinaryArtifactKind(fileSystem: fileSystem, path: path) {
+            return (path, kind)
+        }
+
+        // try to find a matching subdirectory
+        let subdirectories = try fileSystem.getDirectoryContents(path)
+            .map{ path.appending(component: $0) }
+            .filter { fileSystem.isDirectory($0) }
+
+        for subdirectory in subdirectories {
+            if let (path, kind) = try deriveBinaryArtifact(fileSystem: fileSystem, path: subdirectory) {
+                return (path, kind)
+            }
+        }
+
+        return .none
+    }
+
+    private static func deriveBinaryArtifactKind(fileSystem: FileSystem, path: AbsolutePath) throws -> BinaryTarget.Kind? {
+        let files = try fileSystem.getDirectoryContents(path)
+            .map{ path.appending(component: $0) }
+            .filter { fileSystem.isFile($0) }
+
+        if let infoPlist = files.first(where: { $0.basename.lowercased() == "info.plist" }) {
+            let decoder = PropertyListDecoder()
+            let data: Data = try fileSystem.readFileContents(infoPlist)
+            if let _ = try? decoder.decode(XCFrameworkMetadata.self, from: data) {
+                return .xcframework
+            }
+        }
+
+        if let infoJSON = files.first(where: { $0.basename.lowercased() == "info.json" }) {
+            if let _ = try? ArtifactsArchiveMetadata.parse(fileSystem: fileSystem, rootPath: infoJSON.parentDirectory) {
```

Comment:
Same comment as with Info.plist, should we emit the potential error here somewhere?